src/main/java/com/redhat/rhjmc/containerjfr/commands/internal/UploadRecordingCommand.java

```diff
@@ -179,17 +170,28 @@ public boolean validate(String[] args) {
 
         boolean isValidTargetId = validateTargetId(targetId);
         if (!isValidTargetId) {
-            cw.println(String.format("%s is an invalid connection specifier", args[0]));
+            cw.println(String.format("%s is an invalid connection specifier", targetId));
         }
 
         boolean isValidRecordingName = validateRecordingName(recordingName);
         if (!isValidRecordingName) {
             cw.println(String.format("%s is an invalid recording name", recordingName));
         }
 
-        // TODO validate datasource URL
+        boolean isValidDatasourceUrl = true;
+        try {
+            new java.net.URL(env.getEnv(GRAFANA_DATASOURCE_ENV));
+        } catch (MalformedURLException e) {
+            cw.println(String.format("%s is an invalid datasource URL", GRAFANA_DATASOURCE_ENV));
+            isValidDatasourceUrl = false;
+        }
 
-        return isValidTargetId && isValidRecordingName;
+        return isValidTargetId && isValidRecordingName && isValidDatasourceUrl;
+    }
+
+    @Override
+    public boolean isAvailable() {
+        return env.hasEnv(GRAFANA_DATASOURCE_ENV);
```

Comment:
Just to be safe, this should also call through to `super.isAvailable()` and `&&` the results together.